File: solutions/graphs/207-course-schedule.py
I'll help you create a solution for the Course Schedule problem. I'll implement it with clear comments and proper structure.

```python
#!/usr/bin/env python3
"""
Course Schedule Implementation - LeetCode 207

This module provides a solution to determine if it's possible to finish all courses
given their prerequisites. It uses a directed graph approach with cycle detection.

Problem:
There are numCourses courses you have to take, labeled from 0 to numCourses - 1.
Some courses have prerequisites. Prerequisites are given as pairs [a, b],
which means you must take course b before course a.

Return true if you can finish all courses. Otherwise, return false.
"""

from typing import List
from collections import defaultdict


class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        """
        Determines if it's possible to finish all courses given the prerequisites.
        
        Args:
            numCourses (int): Total number of courses
            prerequisites (List[List[int]]): List of prerequisite pairs [course, prereq]
        
        Returns:
            bool: True if all courses can be completed, False otherwise
        
        Time Complexity: O(V + E) where V is number of courses and E is number of prerequisites
        Space Complexity: O(V + E)
        """
        # Create adjacency list representation of the graph
        graph = defaultdict(list)
        for course, prereq in prerequisites:
            graph[course].append(prereq)
        
        # Set to keep track of visited nodes in current DFS path
        path = set()
        # Set to keep track of all visited nodes
        visited = set()
        
        def has_cycle(course: int) -> bool:
            """
            Detects if there's a cycle in the course dependencies using DFS.
            
            Args:
                course (int): Current course being checked
            
            Returns:
                bool: True if cycle is detected, False otherwise
            """
            # If course is in current path, we found a cycle
            if course in path:
                return True
            # If course was already visited and verified to be safe
            if course in visited:
                return False
            
            # Add course to current path
            path.add(course)
            
            # Check all prerequisites of current course
            for prereq in graph[course]:
                if has_cycle(prereq):
                    return True
            
            # Remove course from current path (backtracking)
            path.remove(course)
            # Mark course as fully visited
            visited.add(course)
            
            return False
        
        # Check each course for cycles
        for course in range(numCourses):
            if has_cycle(course):
                return False
        
        return True


def test_solution():
    """
    Test cases for the course schedule implementation.
    """
    solution = Solution()
    
    # Test case 1: Simple case with no cycle
    assert solution.canFinish(2, [[1,0]]) == True
    
    # Test case 2: Case with cycle
    assert solution.canFinish(2, [[1,0], [0,1]]) == False
    
    # Test case 3: More complex case without cycle
    assert solution.canFinish(4, [[1,0], [2,1], [3,2]]) == True
    
    # Test case 4: Empty prerequisites
    assert solution.canFinish(2, []) == True
    
    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()
```

This implementation provides a solution to the Course Schedule problem with the following features:

1. Clear documentation and comments explaining the implementation
2. Proper type hints using the typing module
3. Efficient solution using depth-first search (DFS) for cycle detection
4. Test cases to verify the implementation
5. Proper handling of edge cases
6. Following Python conventions and best practices

The solution uses a graph-based approach where:
- Courses are represented as nodes
- Prerequisites are represented as directed edges
- A cycle in the graph means it's impossible to complete all courses
- The absence of cycles means all courses can be completed

The main algorithm uses DFS with cycle detection to determine if the course schedule is possible. It maintains two sets:
- `path`: tracks nodes in the current DFS path (for cycle detection)
- `visited`: tracks all visited nodes (for optimization)

The time complexity is O(V + E) where V is the number of courses and E is the number of prerequisites.
The space complexity is O(V + E) for storing the graph and recursion stack.

File: solutions/graphs/alternatives/207-course-schedule.js
/**
 * 207. Course Schedule
 * Medium
 *
 * Course Schedule - Determines if it's possible to finish all courses Problem: Given numCourses and prerequisites array where prerequisites[i] = [ai, bi] indicates that you must take course bi first if you want to take course ai, return true if you can finish all courses. Otherwise, return false. @param {number} numCourses - The total number of courses @param {number[][]} prerequisites - Array of prerequisite pairs @return {boolean} - Whether it's possible to finish all courses
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * The key insight for solving Course Schedule is to understand the core problem pattern
 * and apply the most efficient algorithmic approach.
 *
 * ### APPROACH:
 * 1. Analyze the problem requirements
 * 2. Choose the optimal data structure
 * 3. Implement the solution step by step
 * 4. Handle edge cases appropriately
 *
 * ### WHY THIS WORKS:
 * This approach works because it leverages the fundamental properties of the problem
 * to achieve an efficient solution.
 *
 * ### EXAMPLE WALKTHROUGH:
 * For a typical input, the algorithm processes the data systematically
 * to produce the expected output.
 *
 * </details>
 */

/**
 * Course Schedule - Determines if it's possible to finish all courses
 * 
 * Problem: Given numCourses and prerequisites array where prerequisites[i] = [ai, bi]
 * indicates that you must take course bi first if you want to take course ai,
 * return true if you can finish all courses. Otherwise, return false.
 * 
 * @param {number} numCourses - The total number of courses
 * @param {number[][]} prerequisites - Array of prerequisite pairs
 * @return {boolean} - Whether it's possible to finish all courses
 */

/**
 * Main function to check if course schedule is possible
 */
function canFinish(numCourses, prerequisites) {
    // Edge cases
    if (numCourses === 0) return true;
    if (!prerequisites || prerequisites.length === 0) return true;

    // Create adjacency list representation of the graph
    const graph = buildGraph(numCourses, prerequisites);
    
    // Array to track visited nodes in current DFS path
    const visited = new Set();
    // Array to track nodes that have been fully explored
    const explored = new Set();

    // Check for cycles starting from each course
    for (let course = 0; course < numCourses; course++) {
        if (!explored.has(course) && hasCycle(course, graph, visited, explored)) {
            return false;
        }
    }

    return true;
}

/**
 * Builds an adjacency list representation of the course prerequisites
 */
function buildGraph(numCourses, prerequisites) {
    const graph = Array(numCourses).fill().map(() => []);
    
    for (const [course, prereq] of prerequisites) {
        graph[course].push(prereq);
    }
    
    return graph;
}

/**
 * Detects if there's a cycle in the graph using DFS
 */
function hasCycle(course, graph, visited, explored) {
    // If we've seen this course in current path, we have a cycle
    if (visited.has(course)) {
        return true;
    }
    
    // If we've already explored this course fully, no need to check again
    if (explored.has(course)) {
        return false;
    }

    // Mark course as visited in current path
    visited.add(course);

    // Check all prerequisites of current course
    for (const prereq of graph[course]) {
        if (hasCycle(prereq, graph, visited, explored)) {
            return true;
        }
    }

    // Remove course from current path as we backtrack
    visited.delete(course);
    // Mark course as fully explored
    explored.add(course);

    return false;
}

// Test cases
function runTests() {
    console.log('Running test cases...');
    
    // Test Case 1: Simple valid case
    console.log(canFinish(2, [[1,0]]));  // Expected: true
    
    // Test Case 2: Cycle detection
    console.log(canFinish(2, [[1,0],[0,1]]));  // Expected: false
    
    // Test Case 3: Multiple prerequisites
    console.log(canFinish(4, [[1,0],[2,1],[3,2]]));  // Expected: true
    
    // Test Case 4: No prerequisites
    console.log(canFinish(2, []));  // Expected: true
    
    // Test Case 5: Complex case with cycle
    console.log(canFinish(4, [[0,1],[1,2],[2,3],[3,1]]));  // Expected: false
}

// Export for use in other modules
module.exports = {
    canFinish,
    buildGraph,
    hasCycle
};

// Uncomment to run tests
// runTests();